Icon Search

aaronmgdr opened the pull request:
### Description

Adds searching to the icons page

### Tested

wrote an interaction test, and played with real search a lot. 

### normal
<img width="1095" alt="Screen Shot 2020-04-07 at 11 16 43 AM" src="https://user-images.githubusercontent.com/3814795/78704840-64008b80-78c1-11ea-927d-6288c1abe850.png">

### search found
<img width="1072" alt="Screen Shot 2020-04-07 at 11 17 00 AM" src="https://user-images.githubusercontent.com/3814795/78704846-6662e580-78c1-11ea-9724-aa5e4c340971.png">

### search none found
<img width="1039" alt="Screen Shot 2020-04-07 at 11 17 16 AM" src="https://user-images.githubusercontent.com/3814795/78704851-66fb7c00-78c1-11ea-8c48-4bd32d3564a2.png">


### Other changes

add small fuzzy-search library. 

Icons and KeyImagery Pages, now fetch assets via getInitalProps rather than thru FetchHoc.

### Related issues

- Fixes #3035 

### Backwards compatibility

new feature

Commit:
Search on Icons Page UI

Commit:
search box input changes on type

Commit:
add fuzzy search and use in basic way

Commit:
sometimes does some off stuff with rendeing but debounce is working nicely

Commit:
WIP for icon search DISable fade in on showcase as fading in and out while searching looks jenky

Commit:
fix tests

Commit:
memoize items, add id

Commit:
add tags to assets

Commit:
whoops

Commit:
add indidcation of icons found

Commit:
dont rerender icons, just change display state as its quicker to render that way

Commit:
pretty pretty fast fast

Commit:
adds interaction tests

Commit:
fetching of tags and assets at the same time

aaronmgdr commented on packages/web/server/AssetBase.ts:
```diff
@@ -38,10 +62,9 @@ async function fetchAssets(sheet: AssetSheet) {
       sort: [{ field: 'Order', direction: 'asc' }],
     })
     .eachPage((records, fetchNextPage) => {
-      records.forEach((r) => assets.push(normalize(r.fields)))
```
fetch just fetches now, data reformating is done separately

aaronmgdr commented on packages/web/src/brandkit/KeyImagery.tsx:
```diff
@@ -134,7 +134,7 @@ const AbstractGraphics = React.memo(
                     key={illo.name}
                     description={illo.description}
                     name={illo.name}
-                    preview={{ uri: illo.preview }}
```
dont create an object on each render

Commit:
Merge branch 'master' into aaronmgdr/icon-search

Commit:
lock it up

Commit:
fetch illos and graphics in get inital props
add back fade in.
fix ts-ignore
update test

Commit:
quicker fade, and protect against stale data

Commit:
fix typo

Commit:
Tags is undefined when there are no tags according to airtable

Commit:
add test for conversion

jmrossy commented on packages/web/server/AssetBase.ts:
```diff
@@ -1,5 +1,6 @@
 import { Attachment, FieldSet, Table } from 'airtable'
 import getConfig from 'next/config'
+import AssetProps from 'src/../fullstack/AssetProps'
```
Why this absolute and relative import?

aaronmgdr replied:
I should really move fullstack folder inside of the src folder.

aaronmgdr replied:
ill create a PR that does this

jmrossy commented on packages/web/src/brandkit/Explorer.tsx:
```diff
@@ -0,0 +1,87 @@
+})
+
+function useVisibleIconIDs(query: string, initial: IconData[]): Set<string> {
+  const [results, setResult] = React.useState(React.useMemo(() => toIDSet(initial), [initial]))
```
Can you unpack this line for me? Not sure what's going on

aaronmgdr replied:
argument passed into useState Becomes its initial state, since I want State to always be a Set of Visible Icons Ids the initial State must be the set of ALL icon ids. hense the `toIDSet(initial)`. 

But we only need to do this first time its run, so useMemo, 

looking at it now I think there is a simpler better way

jmrossy commented on packages/web/src/brandkit/IconsPage.tsx:
```diff
@@ -1,81 +1,64 @@
+
+IconPage.getInitialProps = async ({ req }) => {
+  let icons = []
+  if (req) {
```
what is req?

aaronmgdr replied:
excellent question! 

`req` is request like in express, basically its how to tell we running this serverside or client side.  which is important since if client side we will need to send a request to our endpoint, and if serverside we want to call the function directly. But we want to NOT import that function on client to save our bundle size (and it wont work) so we do an awaited import

aaronmgdr replied:
if req exists its being run on a server, if not its being run in browser

jmrossy commented on packages/web/src/brandkit/KeyImagery.tsx:
```diff
@@ -1,56 +1,82 @@
+
+KeyImageryWrapped.getInitialProps = async ({ req }): Promise<Props> => {
+  let results
+  if (req) {
```
ditto here, not understanding what's going on in this if/else

jmrossy commented on packages/web/src/brandkit/KeyImagery.tsx:
```diff
@@ -1,56 +1,82 @@
-    )
-  })
-)
+const Overview = React.memo(function _Overview() {
```
Are we overusing React.memo maybe?

aaronmgdr replied:
might have gone overboard

jmrossy commented on packages/web/src/brandkit/common/Showcase.tsx:
```diff
@@ -29,6 +30,11 @@ export default React.memo(function Showcase({
+    name,
+    assetType,
+  ])
+  const titleStyle = React.useMemo(() => [fonts.h6, styles.title], [styles.title])
```
ditto, overusing useMemo?

aaronmgdr replied:
yeah probably aye.

jmrossy reviewed (commented):
Mostly looks good, just some questions

Commit:
remove some not so useful memos
add comments explaining  req
less clever, more js idomatic way to only run toIDSet once on initial.

mergify[bot] committed:
Merge branch 'master' into aaronmgdr/icon-search

Commit:
fix type lints

mergify[bot] committed:
Merge branch 'master' into aaronmgdr/icon-search

aaronmgdr committed:
Merge branch 'master' into aaronmgdr/icon-search

aaronmgdr committed:
Merge branch 'master' into aaronmgdr/icon-search
